subsys/mctp/mctp_uart.c

```diff
@@ -0,0 +1,331 @@
+/*
+ * Copyright (c) 2024 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+#include <zephyr/sys/__assert.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/mctp/mctp_uart.h>
+#include <crc-16-ccitt.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(mctp_uart, CONFIG_MCTP_LOG_LEVEL);
+
+#define MCTP_UART_REVISION     0x01
+#define MCTP_UART_FRAMING_FLAG 0x7e
+#define MCTP_UART_ESCAPE       0x7d
+
+const char *UART_EVENT_STRING[] = {
+	"TX Done",     "TX Aborted", "RX Ready", "RX Buffer Request", "RX Buffer Released",
+	"RX Disabled", "RX Stopped",
+};
+
+const char *MCTP_STATE_STRING[] = {
+	"Wait: Sync Start", "Wait: Revision", "Wait: Len",  "Data",
+	"Data: Escaped",    "Wait: FCS1",     "Wait: FCS2", "Wait: Sync End",
+};
+
+struct mctp_serial_header {
+	uint8_t flag;
+	uint8_t revision;
+	uint8_t len;
+};
+
+struct mctp_serial_trailer {
+	uint8_t fcs_msb;
+	uint8_t fcs_lsb;
+	uint8_t flag;
+};
+
+static inline struct mctp_binding_uart *binding_to_uart(struct mctp_binding *b)
+{
+	return (struct mctp_binding_uart *)b;
+}
+
+static void mctp_uart_finish_pkt(struct mctp_binding_uart *uart, bool valid)
+{
+	struct mctp_pktbuf *pkt = uart->rx_pkt;
+
+	__ASSERT_NO_MSG(pkt);
```

Comment:
suggestion: move this ASSERT into the `if (valid)` statement.  No need to shutdown the whole system if it was already going to be a no-op anyways.